code/game/objects/structures/curtains.dm

```diff
@@ -48,16 +149,27 @@
 	opacity = TRUE
 	density = FALSE
 	anchored = TRUE
-
-	var/obj/item/curtain/holder = /obj/item/curtain
+	var/curtain_kind_path = /decl/curtain_kind
 
 /obj/structure/curtain/open
 	icon_state = "open"
 	layer = ABOVE_HUMAN_LAYER
 	opacity = FALSE
 
-/obj/structure/curtain/Initialize()
+/obj/structure/curtain/Initialize(ml, _mat, _reinf_mat, var/decl/curtain_kind/kind = null)
+	if(kind)
+		if(istype(kind))
+			curtain_kind_path = kind.type
+		else if(ispath(kind))
+			curtain_kind_path = kind
+	else
+		kind = GET_DECL(curtain_kind_path)
+	_mat = kind.material_key
+	matter = atom_info_repository.get_matter_for(/obj/structure/curtain, kind.material_key)
```

Comment:
This is an infinite loop - needs to be in/obj/item/curtain/Initialize(), in this instance it's causing an infinite loop of Initialize.

Reply:
I'm not sure I understand what you mean here? Does the atom_info_repository instantiate atoms?

Reply:
Also, if this creates an infinite loop putting it in the item would just create an infinite loop too no?